DAS-2070 - Implement GitHub workflows CI/CD.

owenlittlejohns opened the pull request:
## Description

This PR implements CI/CD to run the `unittest` suite for all PRs. It also adds a Docker image publication workflow for when changes are merged into the `main` branch containing an update to `docker/service_version.txt`.

The workflows are essentially copies of those used in the [HOSS](https://gtihub.com/nasa/harmony-opendap-subsetter) and [Swath Projector](https://github.com/harmony-swath-projector) repositories.

## Jira Issue ID

DAS-2070

## Local Test Steps

It's probably good to check the `bin/build-image`, `bin/build-test` and `bin/run-test` scripts all work on your local machine (given the change in image name).

You can also test this all works with Harmony locally:

* Clone this repository.
* Run `bin/build-image`.
* Update your local Harmony instance to use the new HyBIG image name:
  * In `packages/util/env-defaults` update `HYBIG_SERVICE_QUEUE_URLS` to use an image name of `ghcr.io/nasa/harmony-browse-image-generator:latest`. (It should just be a case of replacing `sds` with `ghcr.io/nasa`)
  * In `services/harmony/env-defaults` update the HyBIG image: `HYBIG_IMAGE=ghcr.io/nasa/harmony-browse-image-generator:latest`.
* Rebuild the image Harmony uses locally with `npm run build-all`.
* Start a new Harmony instance with `bin/bootstrap-harmony`.
* Run through the [HyBIG regression test suite](https://github.com/nasa/harmony-regression-tests/tree/main/test/hybig) using a Jupyter notebook server in the browser, and setting `HARMONY_HOST_URL='http:localhost:3000'`. The tests should run and complete.

Things to look for on this PR itself:

* The unit tests should run and complete successfully.
* I _think_ a Snyk check should also have been kicked off and pass. (Checking both security and licenses)

## PR Acceptance Checklist
* [x] Jira ticket acceptance criteria met.
* ~~`CHANGELOG.md` updated to include high level summary of PR changes.~~ (Already prepared for v1.0.0)
* ~~`docker/service_version.txt` updated if publishing a release.~~ (Already at 1.0.0)
* ~~Tests added/updated and passing.~~ (No new tests needed)
* [x] Documentation updated (if needed).

lyonthefrog commented:
Take out the evasive sds's: 
<img width="1078" alt="image" src="https://github.com/nasa/harmony-browse-image-generator/assets/82548520/935db19c-f0e0-40e6-9a81-610ef8495f0b">

owenlittlejohns commented:
@lyonthefrog - good catches. I made some changes in the [latest commit](https://github.com/nasa/harmony-browse-image-generator/pull/1/commits/8526320bc9bf8e57550547c9e124135369d441ed) to fix most of those:

* Updated all the references in the `exceptions.py` module. I made that just "harmony-browse-image-generator" and made it a variable, so it only ever needs to be tweaked once.
* Updated the references in `build-test` and `build-image`.
* Updated the references in the `service.Dockerfile` and `tests.Dockerfile`.
* Left things as they were in the `CHANGELOG.md`, as that is trying to say that v1.0.0 is the same as `v0.0.11` of the old sds-related Docker image.

lyonthefrog commented:
The HyBIG regression tests were successful, as were the unit tests (report below). And it appears that the Snyk checks below have passed. 

<img width="647" alt="image" src="https://github.com/nasa/harmony-browse-image-generator/assets/82548520/c1f5de04-a703-4f66-8c20-37d7a8bd4984">

lyonthefrog reviewed (approved):
I have already commented on my successful tests and one issue that has already been resolved (sds's), so I approve this.

flamingbear commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,16 @@
+A short description of the changes in this PR.
+
+## Jira Issue ID
+
```
```suggestion
[DAS-XXX](https://bugs.earthdata.nasa.gov/browse/DAS-XXX)
```
What about that, make it a little easier?

owenlittlejohns replied:
[Sounds good](https://github.com/nasa/harmony-browse-image-generator/pull/1/commits/2262f4ff074d985d3db397f045f462c9cecd1ba9)

flamingbear commented on README.md:
```diff
@@ -99,11 +105,36 @@ updated:
-Initially service Docker images will be hosted in AWS Elastic Container
-Registry (ECR). When this repository is migrated to the NASA GitHub
-organisation, service images will be published to ghcr.io, instead.
+The CI/CD for HyBIG is contained in GitHub workslofws in the
```
```suggestion
The CI/CD for HyBIG is contained in GitHub worksflows in the
```

flamingbear commented on README.md:
```diff
@@ -112,3 +143,11 @@ that service version can be released to a Harmony environment by updating the
+You can reach out to the maintainers of this repository via email:
+
+* david.p.auty@nasa.gov
+* matthew.savioe@colorado.edu
```
```suggestion
* matthew.savoie@colorado.edu
```
I don't love email, but I should be able to ignore it.

owenlittlejohns replied:
Gah! Sorry!

flamingbear commented on bin/extract-release-notes.sh:
```diff
@@ -0,0 +1,24 @@
+# 2024-01-23: Copied from Swath Projector repository to HyBIG.
+#
+###############################################################################
+
+CHANGELOG_FILE="CHANGELOG.md"
+VERSION_PATTERN="^## v"
+# Count number of versions in version file:
+number_of_versions=$(grep -c "${VERSION_PATTERN}" ${CHANGELOG_FILE})
+
+if [ ${number_of_versions} -gt 1 ]
+then
+	grep -B 9999 -m 2 "${VERSION_PATTERN}" ${CHANGELOG_FILE} | sed '$d' | sed '$d'
+else
+	cat ${CHANGELOG_FILE}
+fi
```
```suggestion

# File name
CHANGELOG_FILE="CHANGELOG.md"

# Regular expression
VERSION_PATTERN="^## v"

# Read the file and extract text between the first two occurrences of the VERSION_PATTERN
result=$(awk "/$VERSION_PATTERN/{c++; if(c==2) exit;} c==1" "$CHANGELOG_FILE")

# Print the result
echo "$result" |  grep -v "^#"
```
What do you think?  I suppose you'd have to update the comment too.

flamingbear commented on bin/extract-release-notes.sh:
```diff
@@ -0,0 +1,24 @@
+# 2023-06-16: Created.
+# 2023-10-10: Copied from earthdata-varinfo repository to HOSS.
+# 2024-01-03: Copied from HOSS repository to the Swath Projector.
+# 2024-01-23: Copied from Swath Projector repository to HyBIG.
```
```suggestion
# 2024-01-23: Copied and modified from Swath Projector repository to HyBIG.
```

flamingbear reviewed (changes requested):
I found some typos that need to be fixed, a suggestion for extracting the changelog differently.  But overall, this is looking great.

owenlittlejohns reviewed (commented):
@flamingbear - I think I fixed up all the typos and adopted all the changes you suggested. Thanks for catching all that.

flamingbear reviewed (approved):
:shipit: LGTM. Thanks for the updates
